File: 15_caterpillar_method/count_triangles.py
def solution(a):
  n = len(a)
  a.sort()
  rt = 0
  for i in range(n - 2):
    k = i + 2;
    for j in range(i + 1, n - 1):
      while k < n and a[k] < a[i] + a[j]:
        k += 1
      rt += k - j - 1
  return rt

File: 15_caterpillar_method/count_triangles.cpp
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> &a) {
  int n = a.size();
  sort(a.begin(), a.end());
  int rt = 0;
  int k;
  for (int i = 0; i < n - 2; i++) {
    k = i + 2;
    for (int j = i + 1; j < n - 1; j++) {
      while(k < n && a[k] < a[i] + a[j]) {
        k++;
      }
      rt += k - j - 1;
    }
  }
  return rt;
}
